File: 2020/day2/1.rb
def validate(line)
    validation, pw = line.split(':')
    range, letter = validation.split(' ')
    min, max = range.split('-').map(&:to_i)

    return pw.count(letter) >= min && pw.count(letter) <= max 
end

def check_pos(line)
    validation, pw = line.split(':')
    range, letter = validation.split(' ')
    pos1, pos2 = range.split('-').map(&:to_i)

    return (pw[pos1] == letter) ^ (pw[pos2] == letter)
end


data = File.readlines('data.txt', chomp:true)

valid1 = 0
valid2 = 0
for line in data
    if validate line 
        valid1 += 1
    end
    if check_pos line
        valid2 += 1
    end
end

puts "Task 1:"
p valid1

puts "Task 2:"
p valid2

File: day14/1.py
import re
from collections import defaultdict

def def_value():
    return 0

pairs = defaultdict(def_value)
letters = defaultdict(def_value)
instructions = {}

with open('data.txt') as f:    
    for line in f:
        if len(line.strip()) <= 0: #Skip the empty line
            continue
        
        ins = line.strip().split(' -> ')

        if len(ins) > 1:
            instructions[ins[0]] = (ins[0][0] + ins[1], ins[1] + ins[0][1])      
        else:
            for letter in ins[0]:
                letters[letter] += 1
            split = re.findall('..', ins[0] + ins[0][1:])
            for pair in split:
                pairs[pair] += 1
            
for i in range(0, 10):
    newpairs = defaultdict(def_value)
    for key, value in pairs.items():
        pair1, pair2 = instructions[key] #Fetch the new pairs
        newpairs[pair1] += 1 * value
        newpairs[pair2] += 1 * value
        letters[pair1[1]] += 1 * value #Only letter to increase is the middle one
    pairs = newpairs

print('Task 1:')
print(max(letters.values()) - min(letters.values()))

for i in range(10, 40):
    newpairs = defaultdict(def_value)
    for key, value in pairs.items():
        pair1, pair2 = instructions[key] #Fetch the new pairs
        newpairs[pair1] += 1 * value
        newpairs[pair2] += 1 * value
        letters[pair1[1]] += 1 * value #Only letter to increase is the middle one
    pairs = newpairs

print('Task 2:')
print(max(letters.values()) - min(letters.values()))






